Cooling Panel UA Calculation Bug for Large Capacity/Low Water Flow Rate

RKStrand opened the pull request:
Pull request overview
---------------------
This pull request includes changes to fix the defect User file with large ZoneHVAC:CoolingPanel:RadiantConvective:Water crashes in UA calculation (Issue #6098).  In the defect input file, the user had entered a large unit capacity but used a fairly small flow rate.  The result was that the panel would have a better than 100% efficiency for heat transfer and led to a hard crash of EnergyPlus.  The fix traps this and lets the user know how to correct the situation.  In addition to a new unit test, the I/O Reference was modified to clarify the relationship between different input parameters and also their intent.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ x ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ x ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue #6098 
 
### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Code style (parentheses padding, variable names)
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check -- verify this
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add rules to spreadsheet
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Documentation changes in place
 - [ ] Changed docs build successfully
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

RKStrand committed:
Fix for bug and modifications to documentation

This commit corrects the bug by checking for the condition that caused
the crash and gracefully exiting out with a severe/fatal error and
helpful hints as to how to fix the problem.  The commit also contains
corrections to the documentation to better describe the input fields
and how to avoid the problem that was discovered in the code.

RKStrand committed:
Addition of unit test for bug fix

Addition of a unit test for this bug fix to verify that the fix is
working properly and the bad cases will produce an error message rather
than crash the program.

jasondegraw commented on doc/input-output-reference/src/overview/group-radiative-convective-units.tex:
```diff
@@ -571,15 +571,15 @@ \subsubsection{Inputs}\label{inputs-3-028-1}
 \paragraph{Field: Rated Water Mass Flow Rate}\label{field-rated-water-mass-flow-rate}
 
-This field is the rated standard water flow rate in kg/s that was used to obtain the rated capacity under rated conditions.  This value along with the other rated conditions is used to calculate the overall heat transfer coefficient (U-Value) that will be used to characterize the heat transfer between the water being circulated through the panel and the air in the space.  The default value is 0.0 ...
+This field is the rated standard water flow rate in kg/s that was used to obtain the design cooling capacity under rated temperature conditions.  This value along with the rated space and water inlet temperatures as well as the cooling capacity (whether autosized or entered by the user) to calculate the overall heat transfer coefficient (U-Value) that will be used to characterize the heat transfe ...
```
I think that perhaps the second sentence here needs a "is used" someplace.

RKStrand replied:
@jasondegraw I agree...I thought I had commented on this earlier and it is not showing up here now.  Anyway, yes, I will correct this as soon as possible.

jasondegraw commented on src/EnergyPlus/ChilledCeilingPanelSimple.cc:
```diff
@@ -1015,29 +1015,62 @@ namespace CoolingPanelSimple {
+	void
+	SizeCoolingPanelUA(
+					 int const CoolingPanelNum,
+					 bool & SizeUAError
```
Could we change this function so rather than returning the error status via the reference it returns true on success and false on failure? It's a more "C++" way of doing this.

RKStrand replied:
@jasondegraw Any objections to lumping this comment together with some other C++ modifications that have been requested?  That is already the subject for Followup for new cooling panel ZoneHVAC:CoolingPanel:RadiantConvective:Water (GitHub Issue #5823).

jasondegraw replied:
@RKStrand That sounds good to me.

jasondegraw replied:
@RKStrand I think maybe I was unclear on what I was suggesting here, sorry if that was the case. What I was suggesting here was to change the function signature to `bool SizeCoolingPanelUA(int const CoolingPanelNum)` rather than use a reference.

jasondegraw reviewed (changes requested):
There are a couple small things that I picked on, but other than that it looks good.

RKStrand committed:
Addressed review comments

Fixed issues noted in review

RKStrand committed:
Merge branch 'develop' into 147520311-CoolingPanelUACalcBug

RKStrand committed:
Correction to logic

Missed a flag assignment that should have been flipped with the switch
from an Error flag to a Success flag.  Really won’t make any difference
but this keeps it in line with the design of the original fix.

RKStrand commented:
@jasondegraw @mjwitte Ok, I decided (after putting additional language in the other defect description) to just go ahead and fix this now.  So, the documentation was altered to fix the missing words and the code was updated as Jason requested.  So, hopefully this is ready to go.  Let me know if there is anything else that needs to be fixed/adjusted.  Thanks!

RKStrand commented:
@jasondegraw @mjwitte Ok, all of the review comments have been included and this is ready for final review.  I did end up correcting the documentation and making the OO change in the flag that Jason requested.  Thanks for your help with this!

RKStrand commented:
Jason:

Sorry for the confusion.  I want to make sure I understand what you are asking before I do anything else because I need to finish this and move on to other things.

If I am understanding you correctly, you want SizeCoolingPanelUA subroutine into a bool function that returns true or false based on the error flag status.  Is that correct?  If not, please clarify further.  If this is correct, then I wonder if that is really the right approach.  This subroutine does calculate a UA value which is stored in the data structure.  I always thought a function is a single purpose entity that cannot really do anything else.  But perhaps that is just my F90 background.  Can you comment on this?

Thanks,

Rick


> On Jul 31, 2017, at 9:53 AM, Jason DeGraw <notifications@github.com> wrote:
> 
> @jasondegraw commented on this pull request.
> 
> In src/EnergyPlus/ChilledCeilingPanelSimple.cc <https://github.com/NREL/EnergyPlus/pull/6181#discussion_r130375649>:
> 
> > @@ -1015,29 +1015,62 @@ namespace CoolingPanelSimple {
>  		
>  		RegisterPlantCompDesignFlow( CoolingPanel( CoolingPanelNum ).WaterInletNode, CoolingPanel( CoolingPanelNum ).WaterVolFlowRateMax );
>  
> +		bool SizeUAErrorFlag = false;
> +		SizeCoolingPanelUA( CoolingPanelNum, SizeUAErrorFlag );
> +		if ( SizeUAErrorFlag ) ShowFatalError( "SizeCoolingPanelUA: Program terminated for previous conditions." );
> +
> +	}
> +
> +	void
> +	SizeCoolingPanelUA(
> +					 int const CoolingPanelNum,
> +					 bool & SizeUAError
> @RKStrand <https://github.com/rkstrand> I think maybe I was unclear on what I was suggesting here, sorry if that was the case. What I was suggesting here was to change the function signature to bool SizeCoolingPanelUA(int const CoolingPanelNum) rather than use a reference.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/NREL/EnergyPlus/pull/6181#discussion_r130375649>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AFiTvXycHLnMyQYfzl5ymTuuaCv1bCw9ks5sTeqGgaJpZM4OAWE8>.
>

jasondegraw commented:
@RKStrand Yes, that is what I am suggesting. It is typical in C++ code to see functions that can "fail" return a boolean status. That doesn't mean the function is doing more than one thing, though. It's just reporting the result of the function call in a better way.

RKStrand committed:
Merge branch 'develop' into 147520311-CoolingPanelUACalcBug

RKStrand commented:
@jasondegraw OK, so even though the "function" is doing other calculations that will be stored in the data structure for the cooling panel, I can (should) define this as a bool function.  I will go ahead and do that and let you know when that is finished.  Hopefully it won't be too long.

RKStrand committed:
More changes due to review

Another series of changes to address review comments from Jason.
Hopefully this corrects the issue that he noted.

RKStrand commented:
@jasondegraw I have committed some additional changes that should work and address the issue (I hope).  It seems to be working so if all the checks pass and you are happy with this, please complete the review and merge.  Thanks!

rraustad commented on src/EnergyPlus/ChilledCeilingPanelSimple.cc:
```diff
@@ -1015,34 +1015,72 @@ namespace CoolingPanelSimple {
+	SizeCoolingPanelUA(
+					 int const CoolingPanelNum
+					 )
+	{
```
spacing severely messed up here

RKStrand replied:
Yes, this was messed up.  Fixed now.

rraustad commented on src/EnergyPlus/ChilledCeilingPanelSimple.cc:
```diff
@@ -1015,34 +1015,72 @@ namespace CoolingPanelSimple {
 		
 	}
-	
+		
```
Added a tab here?

RKStrand replied:
Also deleted the extra tab here and a couple other places in this new routine.

rraustad commented on src/EnergyPlus/ChilledCeilingPanelSimple.cc:
```diff
@@ -1015,34 +1015,72 @@ namespace CoolingPanelSimple {
+		Real64 RatCapToTheoMax; // Ratio of unit capacity to theoretical maximum output based on rated parameters
+
+		SizeCoolingPanelUA = true;
 		Cp = 4120.0; // Just an approximation, don't need to get an exact number
```
Why don't we need an exact number if a few lines below we test for RatCapToTheoMax < 1.0 ?? What if RatCapToTheoMax is 1.0001 because we weren't accurate on Cp?

RKStrand replied:
@rraustad Well, I see your point that perhaps this needs a bit of tolerance in here.  This is mostly to trap the clearly illogical.  I will outline something in response to your next comment.

rraustad commented on src/EnergyPlus/ChilledCeilingPanelSimple.cc:
```diff
@@ -1015,34 +1015,72 @@ namespace CoolingPanelSimple {
 		Tinletr = CoolingPanel( CoolingPanelNum ).RatedWaterTemp;
 		Tzoner = CoolingPanel( CoolingPanelNum ).RatedZoneAirTemp;
+		RatCapToTheoMax = std::abs(Qrated) / ( MDotXCp * std::abs( Tinletr - Tzoner ) );
+		if (RatCapToTheoMax >= 1.0 ) {
```
Why can't RatCapToTheoMax = 1?

RKStrand replied:
@rraustad Two reasons. First because physically you can't get more heat transfer out of the fluid than physically possible.  Having a ratio of higher than 1 means that the exit temperature of the water leaving the panel would be warmer than the room temperature, violating the 2nd Law.  Second because in the code following this it would take the log of a zero or negative number which would cause a crash. Basically the user is setting up a cooling panel that is better than a perfect heat transfer device using rated input that does not make sense.  Admittedly, with a fixed Cp number this could result if the user is "flying to close to the sun".  I will adjust the code to be a little more gracious but also to limit the RatCaptoTheoMax number to something that will still allow a calculation.  So, it will only trap things when there is really a problem with the input.

rraustad commented on src/EnergyPlus/ChilledCeilingPanelSimple.cc:
```diff
@@ -1015,34 +1015,72 @@ namespace CoolingPanelSimple {
+			ShowContinueError( "Check the rated input for temperatures, flow, and capacity for this unit." );
+			ShowContinueError( "The ratio of the capacity to the rated theoretical maximum must be less than unity." );
+			ShowContinueError( "The most likely cause for this is probably either the capacity (whether autosized or hardwired) being too high or the rated flow being too low." );
+			SizeCoolingPanelUA = false;
```
What if the water temp < zone temp? Just saying there could be more information here.

RKStrand replied:
@rraustad I'll add some more text to the error message to help the user more.  By the way, right after this, there is an error check regarding the temperatures but I will add something to the error message you highlighted to discuss temperatures as well.

RKStrand committed:
Merge branch 'develop' into 147520311-CoolingPanelUACalcBug

rraustad commented on doc/input-output-reference/src/overview/group-radiative-convective-units.tex:
```diff
@@ -571,15 +571,15 @@ \subsubsection{Inputs}\label{inputs-3-028-1}
 
-This field is for the cooling panel unit nominal cooling capacity in watts. This field can be autosized by EnergyPlus. When the Cooling Design Capacity Method is \emph{CoolingDesignCapacity}~ and this input is is blank, autosizing is assumed. Design day sizing run must be specified.
+This field is for the cooling panel unit nominal cooling capacity in watts at the rated flow rate of water through the unit (see Rated Water Mass Flow Rate field above). This field can be autosized by EnergyPlus. When the Cooling Design Capacity Method is \emph{CoolingDesignCapacity}~ and this input is is blank, autosizing is assumed. Design day sizing run must be specified.
 
```
if another commit is required, this text could be corrected. "is is"

... and this input is is blank, autosizing is assumed.

RKStrand replied:
@rraustad I see it and will correct it.  Need to make another commit to add things to the code per your other comments so this will get in.

RKStrand committed:
More review changes

Clarified error messages and added additional logic to trap other
potential problems in the UA calculation.  Also modified docs (was
actually a problem in several places where “is is” mistake was made—all
were corrected in this file.

RKStrand committed:
Merge branch 'develop' into 147520311-CoolingPanelUACalcBug

RKStrand commented:
@rraustad Ok, I have another commit that improves based on what you suggested and also fixes the "is is" doc problem (which actually showed up in several places--I did a Find and fixed them all in this file).  I also merged with develop and the test suite is running.  Hopefully this takes care of things and this is now ready to officially merge.  Thanks for reviewing this!

jasondegraw commented:
@Myoldmopar @mjwitte I believe that @RKStrand has addressed all of the comments and this is ready to merge. I'm going to assign you instead of me because this PR has no input changes and can drop in whenever you think is best.

rraustad commented:
Diff's are either DELight, get_bucket - head_bucket, or invalid string position. This looks OK.

rraustad commented:
Reviewed, merged, and closed.
